src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -684,27 +684,29 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                 if ns != Namespace::ValueNS {
                     return None;
                 }
-                debug!("looking for variants or fields named {} for {:?}", item_name, did);
+                debug!("looking for fields named {} for {:?}", item_name, did);
                 // FIXME: this doesn't really belong in `associated_item` (maybe `variant_field` is better?)
-                // NOTE: it's different from variant_field because it resolves fields and variants,
+                // NOTE: it's different from variant_field because it only resolves struct fields,
                 // not variant fields (2 path segments, not 3).
+                // We need to handle struct (and union) fields in this code because syntactically their paths
+                // are identical to associated item paths: `module::Type::field` and `module::Type::Assoc`.
+                // On the other hand, variant fields can't be mistaken for associated items because they look
+                // like this: `module::Type::Variant::field`. I'm not quite sure why we don't handle variant
+                // paths here though, since they also look like associated items: `module::Type::Variant`.
```

Comment:
Variant paths are already handled by `resolve_str_path_error`. We don't need to special case them.

Reply:
Why aren't struct fields handled there as well?

Reply:
I'm going to approve the comment as it is for now and then we can tweak it (and the code) later.

Reply:
`resolve_str_path_error` is a direct call into the resolver. There's no context in which `SomeStruct::some_field` is valid Rust outside of intra-doc links.

Reply:
The whole reason this PR works at all is because enum variants are handled by the resolver.

Reply:
Ah, I see. So does `resolve_str_path_error` not handle associated items?

Reply:
Not in all cases, because associated items can depend on type-based resolution (e.g. `Vec::IntoIter`).